servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1478,8 +1517,180 @@ public void setState(long fileId, SetStateOptions options)
+          if (workerBlockCounts.containsKey(blockLocation.workerId)) {
+            workerBlockCounts.put(blockLocation.workerId,
+                workerBlockCounts.get(blockLocation.workerId) + 1);
+          } else {
+            workerBlockCounts.put(blockLocation.workerId, 1);
+          }
+        }
+      }
+    } catch (FileDoesNotExistException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    } catch (InvalidPathException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    }
+
+    if (workerBlockCounts.size() == 0) {
+      LOG.error("The file " + fileId + " does not exist on any worker");
+      return IdUtils.INVALID_WORKER_ID;
+    } else if (workerBlockCounts.size() > 1) {
+      LOG.info("The file is stored at more than one worker: " + workerBlockCounts);
+    }
+
+    // return the first worker that has all the blocks
+    for (Entry<Long, Integer> entry : workerBlockCounts.entrySet()) {
+      if (entry.getValue() == totalBlockNum) {
+        return entry.getKey();
+      }
+    }
+
+    LOG.error("Not all the blocks of file {} stored on the same worker", fileId);
+    return IdUtils.INVALID_WORKER_ID;
+  }
+
+  /**
+   * Polls the files to send to the given worker for persistence. It also removes files from the
+   * worker entry in {@link #mWorkerToAsyncPersistFiles}.
+   *
+   * @param workerId the worker id
+   * @return the list of files
+   * @throws FileDoesNotExistException if the file does not exist
+   * @throws InvalidPathException if the path is invalid
+   */
+  private List<PersistFile> pollFilesToCheckpoint(long workerId)
       throws FileDoesNotExistException, InvalidPathException {
+    List<PersistFile> filesToPersist = Lists.newArrayList();
+    List<Long> fileIdsToPersist = Lists.newArrayList();
+
+    synchronized (mWorkerToAsyncPersistFiles) {
+      if (!mWorkerToAsyncPersistFiles.containsKey(workerId)) {
+        return filesToPersist;
+      }
+
+      Set<Long> scheduledFiles = mWorkerToAsyncPersistFiles.get(workerId);
+      for (long fileId : scheduledFiles) {
+        InodeFile inode = (InodeFile) mInodeTree.getInodeById(fileId);
+        if (inode.isCompleted()) {
+          fileIdsToPersist.add(fileId);
+          List<Long> blockIds = Lists.newArrayList();
+          for (FileBlockInfo fileBlockInfo : getFileBlockInfoList(fileId)) {
+            blockIds.add(fileBlockInfo.blockInfo.blockId);
+          }
+
+          filesToPersist.add(new PersistFile(fileId, blockIds));
+          // update the inode file persisence state
+          inode.setPersistenceState(PersistenceState.IN_PROGRESS);
+          scheduledFiles.remove(fileId);
+        }
+      }
+    }
+
+    mWorkerToAsyncPersistFiles.get(workerId).remove(fileIdsToPersist);
+
+    // write to journal
+    PersistFilesRequestEntry persistFilesRequest =
```

Comment:
Hrmmm, I'm not completely sure about this, but I think journal writes might have to be serialized, and the way this master has been doing it is with mInodeTree. However, it is possible my assumption is not true.